From ca2d5e4f599d05f8294ad82da8c5c21c98a29df0 Mon Sep 17 00:00:00 2001
From: Ren Jianing <jianing.ren@rock-chips.com>
Date: Fri, 19 Jun 2020 11:07:21 +0800
Subject: [PATCH] usb: dwc3: gadget: fix isoc out transfer issue

When I test the uac gadget function, I find that the uac
gadget out transfer may stop expectedly and the alsa buffer
will be empty which results to input/output error.

In some cases, we will get an endpoint event
DWC3_DEPEVT_XFERINPROGRESS and the event status is
DEPEVT_STATUS_MISSED_ISOC, and the original code try to issue
an EndTransfer command. After the EndTransfer is completed,
we will get a XferNotReady event. However, it will fail to
giveback the req to the udc gadget driver due to an additional
request unmap operation.

To fix this problem, this patch return directly in kick transfer
if there are isochronous endpoints and retry to transfer next time.

Signed-off-by: Ren Jianing <jianing.ren@rock-chips.com>
Change-Id: I80a8e7c0caa9129e5736bbf89909662c2f3ed4fe
---
 drivers/usb/dwc3/gadget.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/drivers/usb/dwc3/gadget.c b/drivers/usb/dwc3/gadget.c
index faf4e35f9996..6cf2ebb88d8e 100644
--- a/drivers/usb/dwc3/gadget.c
+++ b/drivers/usb/dwc3/gadget.c
@@ -1341,13 +1341,12 @@ static int __dwc3_gadget_kick_transfer(struct dwc3_ep *dep)
 	ret = dwc3_send_gadget_ep_cmd(dep, cmd, &params);
 	if (ret < 0) {
 		/*
-		 * Isochronous endpoints in request needs to
+		 * Isochronous endpoints request needs to
 		 * return directly and retry to transfer next
 		 * time. Otherwise, it will fail to giveback
 		 * the req to the udc gadget driver.
 		 */
-		if (usb_endpoint_xfer_isoc(dep->endpoint.desc) &&
-		    usb_endpoint_dir_in(dep->endpoint.desc))
+		if (usb_endpoint_xfer_isoc(dep->endpoint.desc))
 			return ret;
 		/*
 		 * FIXME we need to iterate over the list of requests
-- 
2.35.3

